base/compiler/tfuncs.jl

```diff
@@ -391,15 +394,17 @@ add_tfunc(isa, 2, 2,
                   if t === Bottom
                       return Const(false)
```

Comment:
> though of course Union{} would be a better answer from the tfunc here

Changing this to `return Bottom` seems to fix `g9765()`, and I *think* is correct based on this PR's `instanceof_tfunc` implementation. Unsure, though, this stuff hurts my brain 😛 

I'll run tests and see how it goes.

Reply:
That doesn't look quite right. Here's the patch I'm using over in #27972 which needed this same fix:
```diff
diff --git a/base/compiler/tfuncs.jl b/base/compiler/tfuncs.jl
index 161e26e984..a7b9d5fa58 100644
--- a/base/compiler/tfuncs.jl
+++ b/base/compiler/tfuncs.jl
@@ -388,12 +390,18 @@ add_tfunc(typeassert, 2, 2,
               return typeintersect(v, t)
           end, 4)
 add_tfunc(isa, 2, 2,
-          function (@nospecialize(v), @nospecialize(t))
-              t, isexact = instanceof_tfunc(t)
+          function (@nospecialize(v), @nospecialize(tt))
+              t, isexact = instanceof_tfunc(tt)
+              if t === Bottom
+                  # check if t could be equivalent to typeof(Bottom), since that's valid in `isa`, but the set of `v` is empty
+                  # if `t` cannot have instances, it's also invalid on the RHS of isa
+                  if typeintersect(widenconst(tt), Type) === Union{}
+                      return Union{}
+                  end
+                  return Const(false)
+              end
               if !has_free_typevars(t)
-                  if t === Bottom
-                      return Const(false)
-                  elseif v ⊑ t
+                  if v ⊑ t
                       if isexact
                           return Const(true)
                       end
@@ -401,7 +409,7 @@ add_tfunc(isa, 2, 2,
                       # this tests for knowledge of a leaftype appearing on the LHS
                       # (ensuring the isa is precise)
                       return Const(false)
-                  elseif isexact && typeintersect(v, t) === Bottom
+                  elseif typeintersect(v, t) === Bottom
                       if !iskindtype(v) #= subtyping currently intentionally answers this query incorrectly for kinds =#
                           return Const(false)
                       end
diff --git a/test/compiler/compiler.jl b/test/compiler/compiler.jl
index 9c6d32d529..e22b12f216 100644
--- a/test/compiler/compiler.jl
+++ b/test/compiler/compiler.jl
@@ -1189,7 +1189,7 @@ let isa_tfunc = Core.Compiler.T_FFUNC_VAL[
     @test isa_tfunc(typeof(Union{}), Const(Int)) === Const(false) # any result is ok
     @test isa_tfunc(typeof(Union{}), Const(Union{})) === Const(false)
     @test isa_tfunc(typeof(Union{}), typeof(Union{})) === Const(false)
-    @test isa_tfunc(typeof(Union{}), Union{}) === Const(false) # any result is ok
+    @test isa_tfunc(typeof(Union{}), Union{}) === Union{}
     @test isa_tfunc(typeof(Union{}), Type{typeof(Union{})}) === Const(true)
     @test isa_tfunc(typeof(Union{}), Const(typeof(Union{}))) === Const(true)
     let c = Conditional(Core.SlotNumber(0), Const(Union{}), Const(Union{}))
@@ -1204,7 +1204,7 @@ let isa_tfunc = Core.Compiler.T_FFUNC_VAL[
     @test isa_tfunc(Val{1}, Type{Val{T}} where T) === Bool
     @test isa_tfunc(Val{1}, DataType) === Bool
     @test isa_tfunc(Any, Const(Any)) === Const(true)
-    @test isa_tfunc(Any, Union{}) === Const(false) # any result is ok
+    @test isa_tfunc(Any, Union{}) === Union{}
     @test isa_tfunc(Any, Type{Union{}}) === Const(false)
     @test isa_tfunc(Union{Int64, Float64}, Type{Real}) === Const(true)
     @test isa_tfunc(Union{Int64, Float64}, Type{Integer}) === Bool
```

Reply:
Ah, nice - this fixes the `isa_tfunc(typeof(Union{}), Const(Union{}))` that I broke with my version of the fix (it should indeed return `Const(false)`, but my patch was making it return `Union{}`.

Thanks!